### ISSUE TITLE
Poisson reconstruction in parallel threads segfaults

### ISSUE THREAD

**Issue by nblythe** [PointCloudLibrary/pcl#1845]
## Your Environment
* Operating System and version: Ubuntu 16.10
* Compiler: gcc 6.2
* PCL Version: 1.8 as packaged for Ubuntu

## Expected Behavior
Two threads should be able to create two different instances of pcl::Poisson and run them on two different input clouds at the same time.

## Current Behavior
Segfault

## Code to Reproduce
[http://pastebin.com/yJrpcpTL](http://pastebin.com/yJrpcpTL)

**Comment by nblythe** [comment]
Confirmed this is in trunk too

**Comment by nblythe** [comment]
The problem is surface/include/pcl/surface/3rdparty/poisson4/allocator.h, which isn't thread-safe.

**Comment by nblythe** [comment]
Possible fixes include switching to member instances instead of static instances (will break memory usage reporting) and mutex-locking allocations.

**Comment by nblythe** [comment]
Unfortunately there's no easy fix for this. Switching to member instances isn't straightforward, because SetAllocator() is called out-of-scope in several places. Mutex-locking allocations won't work because Allocator exposes state that implies single-consumer. The best fix would be to write new thread-safe implementation of Allocator.

**Comment by AlrIsmail** [comment]
This is a third-party library problem that has been fixed, so won't it be better to just update it?
